library/core/src/primitive_docs.rs

```diff
@@ -868,6 +868,33 @@ mod prim_array {}
 /// * Further methods that return iterators are [`.split`], [`.splitn`],
 ///   [`.chunks`], [`.windows`] and more.
 ///
+/// ## Use in slice-based dynamically-sized types
+///
+/// Slices can be used to construct "slice-based dynamically-sized types", also
+/// known as "slice DSTs" or "custom DSTs":
+///
+/// ```
+/// struct SliceDst {
+///     a: u8,
+///     b: u16,
+///     c: [u32],
+/// }
+/// ```
+///
+/// Just like a slice, a slice DST is a `!Sized` type, and cannot be used by-value.
+/// References to slice DSTs encode the number of elements in the trailing slice
+/// field:
+///
+/// ```
+/// # struct SliceDst { a: u8, b: u16, c: [u32] }
+/// let pointer_size = std::mem::size_of::<&u8>();
+/// assert_eq!(2 * pointer_size, std::mem::size_of::<&SliceDst>());
+/// ```
+///
+/// Rust guarantees that, if a slice DST compiles successfully, then the instance
+/// of that type with 0 elements in its trailing slice is no more than `isize::MAX`
+/// bytes in size.
```

Comment:
What does "a slice DST compiles successfully" mean? Is there precedent for a statement like this somewhere else that we can follow?

Reply:
It's just meant to account for the fact that you could in principle write down a type which, were it to compile, would violate this guarantee (i.e., a type whose fixed part itself overflows `isize`). I can just remove that parenthetical, though, since it's implicit - obviously you only care about code that compiles.

Reply:
Okay, I see. Do we usually phrase this like that? "Rust guarantees that all types are no larger than `isize`." That's not what I would have said, I would have said that Ruts *requires* that all types fit in an `isize`, and halts compilation otherwise. But I don't know which wording we are using elsewhere.

Do we have a test for this? We have tests for large types but it might be worth having one specifically for this clause.